File: MinjiJu/2022-06-29-longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        
        if(nums.size()==0) return 0;
        
        vector<int> res;
        res.push_back(nums[0]);
        
        for(int i=1; i<nums.size(); i++){
            if(nums[i] > res.back()){
                res.push_back(nums[i]);
            }
            else{
                int j = lower_bound(res.begin(),res.end(),nums[i]) - res.begin();
                res[j] = nums[i];
            }
        }
        return res.size();
    }
};

File: Chris/2022-06-29-LongestIncreasingSubsequence.py
# O(n^2)
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        
        LIS = [1] * len(nums)
        
        for i in range(len(nums)-1,-1,-1):
            for j in range(i+1,len(nums)):
                if(nums[i] < nums[j]):
                    LIS[i] = max(LIS[i], LIS[j] + 1)
        
        return max(LIS)



#O(nlogn)

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        tails = [0] * len(nums)
        size = 0
        
        
        for num in nums:
            l, r = 0, size
            while(l < r):
                m = (l+r)//2
                if(num > tails[m]):
                    l = m + 1
                else:
                    r = m
            
            tails[l] = num
            size = max(size , l +1)
        
        return size